pennylane/optimize/spsa.py

```diff
@@ -0,0 +1,255 @@
+        """Update the parameter array :math:`x` with one step of the optimizer and return
+        the step and the corresponding objective function.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            *args : variable length argument array for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            tuple[list [array], float]: the new variable values :math:`x^{(t+1)}` and the
+            objective function output prior to the step.
+        """
+        g, forward = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(g, args)
+
+        if forward is None:
+            forward = objective_fn(*args, **kwargs)
+
+        # unwrap from list if one argument, cleaner return
+        if len(new_args) == 1:
+            return new_args[0], forward
+        return new_args, forward
+
+    def step(self, objective_fn, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            *args : variable length argument array for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list [array]: the new variable values :math:`x^{(t+1)}`.
+        """
+        g, _ = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(g, args)
+
+        # unwrap from list if one argument, cleaner return
+        if len(new_args) == 1:
+            return new_args[0]
+
+        return new_args
+
+    def increment_k(self):
+        """Increments k"""
+        self.k += 1
+
+    def compute_grad(self, objective_fn, args, kwargs):
+        r"""Compute approximation of gradient of the objective function at the
+        given point.
+
+        Args:
+            objective_fn (function): The objective function for optimization
+            args (tuple): tuple of NumPy array containing the current parameters
+                for objective function
+            kwargs (dict): keyword arguments for the objective function
+
+        Returns:
+            tuple (array): Numpy array containing the gradient
+                :math:`\hat{g}_k(\hat{\theta}_k)` and ``None``
+        """
+        self.increment_k()
+        ck = self.c / (self.k + 1.0) ** self.gamma
+        delta = list(args)
+        thetaplus = list(args)
+        thetaminus = list(args)
+        grad = []
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", False):
+                delta[index] = np.random.choice([-1, 1], size=arg.shape)
+                thetaplus[index] = arg + ck * delta[index]
+                thetaminus[index] = arg - ck * delta[index]
+        yplus = objective_fn(*thetaplus, **kwargs)
+        yminus = objective_fn(*thetaminus, **kwargs)
```

Comment:
What happens if `objective_fn` returns a vector?

Reply:
Is this defined?

Reply:
The linked article considers scalar functions. Nonetheless, there may be extensions to vector-valued functions too.

My question here is more so practical: what happens with the machinery of `qml.SPSAOptimizer` if `objective_fn` *does* return a vector?

It may be a common use case that a user defines a QNode that returns `qml.probs(wires=wires)` for some `wires` and would like to use `qml.SPSAOptimizer`. Does the code break? Is an error being raised? If so, does it provide a descriptive error message? This should be a case tested. :+1:

Reply:
So for now only scalar functions are enough?

Reply:
Yes, that's okay. See that there's a validation check now too. :+1: